pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -204,6 +208,10 @@
 
     private final AtomicReference<ClientCnx> clientCnxUsedForConsumerRegistration = new AtomicReference<>();
     private final List<Throwable> previousExceptions = new CopyOnWriteArrayList<Throwable>();
+    // Key is the ledger id and the entry id, entry is the acker that represents which single messages are acknowledged
+    private final ConcurrentNavigableMap<Pair<Long, Long>, BatchMessageAcker> batchMessageToAcker =
```

Comment:
We should also consider cleaning up the `batchMessageToAcker` 

- After the seek operation
- After the message has been redelivered(Nack or Ack timeout)
- After the consumer reconnects to the broker? (I'm not 100% sure about this part, the cursor-reset might happened on the broker side)

Reply:
I didn't get it. A new map entry could only be only added when a message was acknowledged.

```java
    private MessageIdImpl getMessageIdToAcknowledge(BatchMessageIdImpl messageId, AckType ackType) {
        final BatchMessageAcker acker;
        if (messageId.getAcker() instanceof BatchMessageAckerDisabled) {
            acker = batchMessageToAcker.computeIfAbsent(
```

Why should we clean up the `batchMessageToAcker` after a seek operation?

Regarding the message redelivering or reconnection, if the batch message id was not created by deserialization, the acker inside the `BatchMessageIdImpl` would not be modified. If we only clean up the `batchMessageToAcker` for deserialized batch message id, the behavior would become different.

For example, assuming there are 2 batch message ids (id0 and id1) of the same entry and for the following steps:
1. Acknowledge `id0`
2. Reconnection
3. Acknowledge `id1`

If they are retrieved from the deserialization, the entry will not be acknowledged because the 2nd step cleared the cache, and the `BatchMessageAcker` will be "XO" (`X` represents not acknowledged).

However, if they are just saved from the `message.getMessageId()`, the entry will be acknowledged because the shared `BatchMessageAcker` is "OO" and not affected by the reconnection.

Reply:
The `batchMessageAckTracker` before https://github.com/apache/pulsar/pull/1424 is updated for each message received:

```java
    void receiveIndividualMessagesFromBatch(MessageMetadata msgMetadata, ByteBuf uncompressedPayload,
            MessageIdData messageId, ClientCnx cnx) {
        /* ... */
        batchMessageAckTracker.put(batchMessage, bitSet);
```

But the `batchMessageToAcker` in this PR will be updated only for acknowlegment.

Reply:
> Why should we clean up the batchMessageToAcker after a seek operation?

The seek API will lead the consumer to consume from a new start message ID. It can be an earlier position. 
Suppose you have 3 batch messages.

0:(0,1,2,3), 
1:(0,1,2), 
2:(0,1,2,3,4,5)

The message 0, 1, and 2:(0,2,4) are acknowledged. Then the consumer seeks to the earliest position.
Then the consumer will receive messages 0, 1, and 2 again.
Due to we have 2:(0,2,4) in the `batchMessageToAcker` and If only 2:(1,3,5) is acked after the seek operation.
But the client will ack the whole message 2. 

From a user perspective, it can be a data loss. We should guarantee the at-lease-once semantic after the seek operation.

Reply:
> Regarding the message redelivering or reconnection, if the batch message id was not created by deserialization, the acker inside the BatchMessageIdImpl would not be modified. If we only clean up the batchMessageToAcker for deserialized batch message id, the behavior would be different.

If the user decides to nack the message, they should not continue to ack it. After the message redelivery, the newly received message with a new Acker. But this PR introduced a shared state. It looks like the newly received message with the different acker can also associate with the old ack state if we don't clean up the shared state.

Reply:
> But the client will ack the whole message 2.

However, it's the same behavior with the current master with those MessageId that have internal acker. This PR keeps the same behavior with the BatchMessageIdImpl that has the `BatchMessageAcker`.

Reply:
> However, it's the same behavior with the current master with those MessageId that have internal acker. This PR keeps the same behavior with the BatchMessageIdImpl that has the BatchMessageAcker.

As I understand. You will get the message again, but with a different Acker after the seek operation. Why should the behavior be the same? Let to try to test locally.

Reply:
> You will get the message again, but with a different Acker after the seek operation.

This "correct" behavior requires you to acknowledge the new messages instead of the old message ids. For example, the following code is the "wrong" behavior:

```java
var msgIds = new ArrayList<MessageId>();
for (int i = 0; i < 3; i++) {
    msgIds.add(consumer.receive().getMessageId());
}
consumer.acknowledge(msgIds.get(0));
consumer.acknowledge(msgIds.get(2));
consumer.seek(Message.earliest);
consumer.acknowledge(msgIds.get(1));
```

Reply:
I have tried them locally.

### The first case uses the message instance to ack the message and seeks to the earliest position after acked half of the messages.

```java
 @Test
 public void testAckAfterSeek() throws Exception {
     final String topic = "persistent://prop/use/ns-abc/testAckAfterSeek-" + UUID.randomUUID();

     @Cleanup
     org.apache.pulsar.client.api.Consumer<String> consumer = pulsarClient.newConsumer(Schema.STRING)
             .topic(topic)
             .subscriptionName("test")
             .subscriptionType(SubscriptionType.Shared)
             .subscribe();

     @Cleanup
     Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
             .topic(topic)
             .batchingMaxPublishDelay(Long.MAX_VALUE, TimeUnit.MILLISECONDS)
             .create();

     final int messages = 10;
     for (int i = 0; i < messages; i++) {
         producer.sendAsync("New message - " + i);
     }
     producer.flush();

     // Only ack the first 5 messages
     for (int i = 0; i < 5; i++) {
         Message<String> received = consumer.receive();
         log.info("[x] Received message: {}", received.getValue());
         consumer.acknowledge(received);
     }

     consumer.seek(MessageId.earliest);

     for (int i = 0; i < messages; i++) {
         Message<String> received = consumer.receive();
         log.info("[x] Received message: {}", received.getValue());
         if (i > 4) {
             consumer.acknowledge(received);
         }
     }

     // Waiting for the message ack command send to the broker.
     Thread.sleep(3000);

     PersistentTopicInternalStats stats = admin.topics().getInternalStats(topic);
     log.info("The topic internal stats : {}", Json.pretty(stats));
 }
```

The output:

```
2022-12-30T16:03:49,319 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ServerCnx@1500] - [/127.0.0.1:59957] Created new producer: Producer{topic=PersistentTopic{topic=persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68}, client=/127.0.0.1:59957, producerName=test-0-0, producerId=0}
2022-12-30T16:03:49,325 - INFO  - [pulsar-client-io-37-5:ProducerImpl@1707] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68] [test-0-0] Created producer on cnx [id: 0xbe2e8d16, L:/127.0.0.1:59957 - R:localhost/127.0.0.1:59949]
2022-12-30T16:03:49,401 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 0
2022-12-30T16:03:49,403 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 1
2022-12-30T16:03:49,403 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 2
2022-12-30T16:03:49,403 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 3
2022-12-30T16:03:49,403 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 4
2022-12-30T16:03:49,405 - INFO  - [main:ConsumerImpl@2144] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68][test] Seeking subscription to the message -1:-1:-1
2022-12-30T16:03:49,406 - INFO  - [pulsar-io-6-2:Consumer@385] - Disconnecting consumer: Consumer{subscription=PersistentSubscription{topic=persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68, name=test}, consumerId=0, consumerName=78011, address=/127.0.0.1:59957}
2022-12-30T16:03:49,407 - INFO  - [pulsar-io-6-2:PersistentDispatcherMultipleConsumers@193] - Removed consumer Consumer{subscription=PersistentSubscription{topic=persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68, name=test}, consumerId=0, consumerName=78011, address=/127.0.0.1:59957} with pending 1 acks
2022-12-30T16:03:49,408 - INFO  - [pulsar-io-6-2:PersistentDispatcherMultipleConsumers@200] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68 / test] All consumers removed. Subscription is disconnected
2022-12-30T16:03:49,408 - INFO  - [pulsar-io-6-2:PersistentSubscription@783] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68][test] Successfully disconnected consumers from subscription, proceeding with cursor reset
2022-12-30T16:03:49,410 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ManagedCursorImpl@1227] - [prop/use/ns-abc/persistent/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68] Initiate reset readPosition to 3:-1 on cursor test
2022-12-30T16:03:49,411 - INFO  - [pulsar-client-io-37-5:ClientCnx@796] - [localhost/127.0.0.1:59949] Broker notification of Closed consumer: 0
2022-12-30T16:03:49,411 - INFO  - [pulsar-client-io-37-5:ConnectionHandler@144] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68] [test] Closed connection [id: 0xbe2e8d16, L:/127.0.0.1:59957 - R:localhost/127.0.0.1:59949] -- Will try again in 0.1 s
2022-12-30T16:03:49,415 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:PulsarMockBookKeeper@122] - Creating ledger 7
2022-12-30T16:03:49,425 - INFO  - [bookkeeper-ml-scheduler-OrderedScheduler-0-0:ManagedCursorImpl$30@3049] - [prop/use/ns-abc/persistent/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68] Updated cursor test with ledger id 7 md-position=3:-1 rd-position=3:1
2022-12-30T16:03:49,433 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ManagedCursorImpl$9@1274] - [prop/use/ns-abc/persistent/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68] reset readPosition to 3:-1 before current read readPosition 3:1 on cursor test
2022-12-30T16:03:49,433 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ServerCnx@1770] - [/127.0.0.1:59957] [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68][test] Reset subscription to message id -1:-1
2022-12-30T16:03:49,433 - INFO  - [pulsar-client-io-37-5:ConsumerImpl@2147] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68][test] Successfully reset subscription to the message -1:-1:-1
2022-12-30T16:03:49,512 - INFO  - [pulsar-timer-76-1:ConnectionHandler@148] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68] [test] Reconnecting after timeout
2022-12-30T16:03:49,517 - INFO  - [pulsar-web-32-14:Slf4jRequestLogWriter@62] - 127.0.0.1 - - [30/Dec/2022:16:03:49 +0800] "GET /lookup/v2/destination/persistent/prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68 HTTP/1.1" 200 217 "-" "Pulsar-Java-v2.12.0-SNAPSHOT" 4
2022-12-30T16:03:49,517 - INFO  - [pulsar-client-io-37-3:ConsumerImpl@780] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68][test] Subscribing to topic on cnx [id: 0xbe2e8d16, L:/127.0.0.1:59957 - R:localhost/127.0.0.1:59949], consumerId 0
2022-12-30T16:03:49,518 - INFO  - [pulsar-io-6-2:ServerCnx@1072] - [/127.0.0.1:59957] Subscribing on topic persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68 / test
2022-12-30T16:03:49,530 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ManagedCursorImpl@2456] - [prop/use/ns-abc/persistent/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68-test] Rewind from 3:-1 to 3:0
2022-12-30T16:03:49,530 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ServerCnx@1156] - [/127.0.0.1:59957] Created subscription on topic persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68 / test
2022-12-30T16:03:49,531 - INFO  - [pulsar-client-io-37-5:ConsumerImpl@914] - [persistent://prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68][test] Subscribed to topic on localhost/127.0.0.1:59949 -- consumer: 0
2022-12-30T16:03:49,541 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 0
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 1
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 2
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 3
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 4
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 5
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 6
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 7
2022-12-30T16:03:49,542 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 8
2022-12-30T16:03:49,543 - INFO  - [main:SubscriptionSeekTest@823] - [x] Received message: New message - 9
2022-12-30T16:03:52,575 - INFO  - [pulsar-web-32-16:Slf4jRequestLogWriter@62] - 127.0.0.1 - - [30/Dec/2022:16:03:52 +0800] "GET /admin/persistent/prop/use/ns-abc/testAckAfterSeek-bfe62332-e92e-4aa1-8387-92ea0ffa6e68/internalStats?metadata=false HTTP/1.1" 200 966 "-" "Pulsar-Java-v2.12.0-SNAPSHOT" 20
2022-12-30T16:03:52,596 - INFO  - [main:SubscriptionSeekTest@833] - The topic internal stats : {
  "entriesAddedCounter" : 1,
  "numberOfEntries" : 1,
  "totalSize" : 275,
  "currentLedgerEntries" : 1,
  "currentLedgerSize" : 275,
  "lastLedgerCreatedTimestamp" : "2022-12-30T16:03:49.057+08:00",
  "waitingCursorsCount" : 2,
  "pendingAddEntriesCount" : 0,
  "lastConfirmedEntry" : "3:0",
  "state" : "LedgerOpened",
  "ledgers" : [ {
    "ledgerId" : 3,
    "entries" : 0,
    "size" : 0,
    "offloaded" : false,
    "underReplicated" : false
  } ],
  "cursors" : {
    "test" : {
      "markDeletePosition" : "3:-1",
      "readPosition" : "3:1",
      "waitingReadOp" : true,
      "pendingReadOps" : 0,
      "messagesConsumedCounter" : 0,
      "cursorLedger" : 7,
      "cursorLedgerLastEntry" : 1,
      "individuallyDeletedMessages" : "[]",
      "lastLedgerSwitchTimestamp" : "2022-12-30T16:03:49.184+08:00",
      "state" : "Open",
      "active" : true,
      "numberOfEntriesSinceFirstNotAckedMessage" : 2,
      "totalNonContiguousDeletedMessagesRange" : 0,
      "subscriptionHavePendingRead" : true,
      "subscriptionHavePendingReplayRead" : false,
      "properties" : { }
    }
  },
  "schemaLedgers" : [ ],
  "compactedLedger" : {
    "ledgerId" : -1,
    "entries" : -1,
    "size" : -1,
    "offloaded" : false,
    "underReplicated" : false
  }
}
```

### The second case uses the deserialized message ID to ack the message and seeks to the earliest position after acked half of the messages.

```java
    @Test
    public void testAckAfterSeek() throws Exception {
        final String topic = "persistent://prop/use/ns-abc/testAckAfterSeek-" + UUID.randomUUID();

        @Cleanup
        org.apache.pulsar.client.api.Consumer<String> consumer = pulsarClient.newConsumer(Schema.STRING)
                .topic(topic)
                .subscriptionName("test")
                .subscriptionType(SubscriptionType.Shared)
                .subscribe();

        @Cleanup
        Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
                .topic(topic)
                .batchingMaxPublishDelay(Long.MAX_VALUE, TimeUnit.MILLISECONDS)
                .create();

        final int messages = 10;
        for (int i = 0; i < messages; i++) {
            producer.sendAsync("New message - " + i);
        }
        producer.flush();

        // Only ack the first 5 messages
        for (int i = 0; i < 5; i++) {
            Message<String> received = consumer.receive();
            log.info("[x] Received message: {}", received.getValue());
            MessageId deserialized = MessageId.fromByteArray(received.getMessageId().toByteArray());
            consumer.acknowledge(deserialized);
        }

        consumer.seek(MessageId.earliest);

        for (int i = 0; i < messages; i++) {
            Message<String> received = consumer.receive();
            log.info("[x] Received message: {}", received.getValue());
            if (i > 4) {
                MessageId deserialized = MessageId.fromByteArray(received.getMessageId().toByteArray());
                consumer.acknowledge(deserialized);
            }
        }

        // Waiting for the message ack command send to the broker.
        Thread.sleep(3000);

        PersistentTopicInternalStats stats = admin.topics().getInternalStats(topic);
        log.info("The topic internal stats : {}", Json.pretty(stats));
    }
```

The output:

```
2022-12-30T16:11:32,926 - INFO  - [pulsar-client-io-37-5:ProducerImpl@1707] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633] [test-0-0] Created producer on cnx [id: 0x18c45e74, L:/127.0.0.1:60303 - R:localhost/127.0.0.1:60295]
2022-12-30T16:11:32,996 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 0
2022-12-30T16:11:32,999 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 1
2022-12-30T16:11:32,999 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 2
2022-12-30T16:11:33,000 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 3
2022-12-30T16:11:33,000 - INFO  - [main:SubscriptionSeekTest@815] - [x] Received message: New message - 4
2022-12-30T16:11:33,002 - INFO  - [main:ConsumerImpl@2248] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633][test] Seeking subscription to the message -1:-1:-1
2022-12-30T16:11:33,003 - INFO  - [pulsar-io-6-2:Consumer@385] - Disconnecting consumer: Consumer{subscription=PersistentSubscription{topic=persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633, name=test}, consumerId=0, consumerName=aec48, address=/127.0.0.1:60303}
2022-12-30T16:11:33,005 - INFO  - [pulsar-io-6-2:PersistentDispatcherMultipleConsumers@193] - Removed consumer Consumer{subscription=PersistentSubscription{topic=persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633, name=test}, consumerId=0, consumerName=aec48, address=/127.0.0.1:60303} with pending 1 acks
2022-12-30T16:11:33,006 - INFO  - [pulsar-io-6-2:PersistentDispatcherMultipleConsumers@200] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633 / test] All consumers removed. Subscription is disconnected
2022-12-30T16:11:33,006 - INFO  - [pulsar-io-6-2:PersistentSubscription@783] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633][test] Successfully disconnected consumers from subscription, proceeding with cursor reset
2022-12-30T16:11:33,008 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ManagedCursorImpl@1227] - [prop/use/ns-abc/persistent/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633] Initiate reset readPosition to 3:-1 on cursor test
2022-12-30T16:11:33,009 - INFO  - [pulsar-client-io-37-5:ClientCnx@796] - [localhost/127.0.0.1:60295] Broker notification of Closed consumer: 0
2022-12-30T16:11:33,009 - INFO  - [pulsar-client-io-37-5:ConnectionHandler@144] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633] [test] Closed connection [id: 0x18c45e74, L:/127.0.0.1:60303 - R:localhost/127.0.0.1:60295] -- Will try again in 0.1 s
2022-12-30T16:11:33,014 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:PulsarMockBookKeeper@122] - Creating ledger 7
2022-12-30T16:11:33,025 - INFO  - [bookkeeper-ml-scheduler-OrderedScheduler-11-0:ManagedCursorImpl$30@3049] - [prop/use/ns-abc/persistent/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633] Updated cursor test with ledger id 7 md-position=3:-1 rd-position=3:1
2022-12-30T16:11:33,032 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ManagedCursorImpl$9@1274] - [prop/use/ns-abc/persistent/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633] reset readPosition to 3:-1 before current read readPosition 3:1 on cursor test
2022-12-30T16:11:33,032 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ServerCnx@1770] - [/127.0.0.1:60303] [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633][test] Reset subscription to message id -1:-1
2022-12-30T16:11:33,033 - INFO  - [pulsar-client-io-37-5:ConsumerImpl@2251] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633][test] Successfully reset subscription to the message -1:-1:-1
2022-12-30T16:11:33,111 - INFO  - [pulsar-timer-76-1:ConnectionHandler@148] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633] [test] Reconnecting after timeout
2022-12-30T16:11:33,118 - INFO  - [pulsar-web-32-14:Slf4jRequestLogWriter@62] - 127.0.0.1 - - [30/Dec/2022:16:11:33 +0800] "GET /lookup/v2/destination/persistent/prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633 HTTP/1.1" 200 217 "-" "Pulsar-Java-v2.12.0-SNAPSHOT" 6
2022-12-30T16:11:33,118 - INFO  - [pulsar-client-io-37-3:ConsumerImpl@884] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633][test] Subscribing to topic on cnx [id: 0x18c45e74, L:/127.0.0.1:60303 - R:localhost/127.0.0.1:60295], consumerId 0
2022-12-30T16:11:33,119 - INFO  - [pulsar-io-6-2:ServerCnx@1072] - [/127.0.0.1:60303] Subscribing on topic persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633 / test
2022-12-30T16:11:33,133 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ManagedCursorImpl@2456] - [prop/use/ns-abc/persistent/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633-test] Rewind from 3:-1 to 3:0
2022-12-30T16:11:33,134 - INFO  - [mock-pulsar-bk-OrderedExecutor-0-0:ServerCnx@1156] - [/127.0.0.1:60303] Created subscription on topic persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633 / test
2022-12-30T16:11:33,135 - INFO  - [pulsar-client-io-37-5:ConsumerImpl@1018] - [persistent://prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633][test] Subscribed to topic on localhost/127.0.0.1:60295 -- consumer: 0
2022-12-30T16:11:33,145 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 0
2022-12-30T16:11:33,145 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 1
2022-12-30T16:11:33,145 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 2
2022-12-30T16:11:33,145 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 3
2022-12-30T16:11:33,145 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 4
2022-12-30T16:11:33,146 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 5
2022-12-30T16:11:33,146 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 6
2022-12-30T16:11:33,146 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 7
2022-12-30T16:11:33,147 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 8
2022-12-30T16:11:33,147 - INFO  - [main:SubscriptionSeekTest@824] - [x] Received message: New message - 9
2022-12-30T16:11:36,182 - INFO  - [pulsar-web-32-16:Slf4jRequestLogWriter@62] - 127.0.0.1 - - [30/Dec/2022:16:11:36 +0800] "GET /admin/persistent/prop/use/ns-abc/testAckAfterSeek-667ef06b-e0a6-43d7-8b71-3ab14fb2a633/internalStats?metadata=false HTTP/1.1" 200 965 "-" "Pulsar-Java-v2.12.0-SNAPSHOT" 23
2022-12-30T16:11:36,209 - INFO  - [main:SubscriptionSeekTest@835] - The topic internal stats : {
  "entriesAddedCounter" : 1,
  "numberOfEntries" : 1,
  "totalSize" : 275,
  "currentLedgerEntries" : 1,
  "currentLedgerSize" : 275,
  "lastLedgerCreatedTimestamp" : "2022-12-30T16:11:32.628+08:00",
  "waitingCursorsCount" : 1,
  "pendingAddEntriesCount" : 0,
  "lastConfirmedEntry" : "3:0",
  "state" : "LedgerOpened",
  "ledgers" : [ {
    "ledgerId" : 3,
    "entries" : 0,
    "size" : 0,
    "offloaded" : false,
    "underReplicated" : false
  } ],
  "cursors" : {
    "test" : {
      "markDeletePosition" : "3:0",
      "readPosition" : "3:1",
      "waitingReadOp" : true,
      "pendingReadOps" : 0,
      "messagesConsumedCounter" : 1,
      "cursorLedger" : 7,
      "cursorLedgerLastEntry" : 2,
      "individuallyDeletedMessages" : "[]",
      "lastLedgerSwitchTimestamp" : "2022-12-30T16:11:32.772+08:00",
      "state" : "Open",
      "active" : true,
      "numberOfEntriesSinceFirstNotAckedMessage" : 1,
      "totalNonContiguousDeletedMessagesRange" : 0,
      "subscriptionHavePendingRead" : true,
      "subscriptionHavePendingReplayRead" : false,
      "properties" : { }
    }
  },
  "schemaLedgers" : [ ],
  "compactedLedger" : {
    "ledgerId" : -1,
    "entries" : -1,
    "size" : -1,
    "offloaded" : false,
    "underReplicated" : false
  }
}

```

You will see that the batch will not be acknowledged by using the message instance but be acknowledged by using the deserialized message ID (from the topic internal stats output).  Hope I didn't miss any details 😁

Reply:
@codelipenghui The key point is about how to use the deserialized batch message id. IMO, when you use message id from the deserialization, before processing them completely, you should not do something else. 
1. Get some message IDs
2. Store them in somewhere else
3. Load them and process them

Just like the example here: Just like the example here: https://github.com/apache/pulsar/pull/19031#discussion_r1059282261, the only difference is that message IDs are stored in memory.

Your example works well because after the seek operation, you didn't use the previous message IDs. Instead, you used the new message IDs, though they represent the same positions with the previous message IDs.

Reply:
I think it's a very hacky to acknowledge the rest messages in a batch after seek. Just like the example I mentioned, if you think that's message lost, users can easily simulate the **message lost** like:

```java
    private Consumer<String> subscribe(String topic) throws Exception {
        return pulsarClient.newConsumer(Schema.STRING)
                .topic(topic)
                .subscriptionName("test")
                .subscriptionType(SubscriptionType.Shared)
                .isAckReceiptEnabled(true)
                .subscribe();
    }

    @Test
    public void testAckAfterSeek() throws Exception {
        final String topic = "testAckAfterSeek";

        Consumer<String> consumer = subscribe(topic);

        @Cleanup
        Producer<String> producer = pulsarClient.newProducer(Schema.STRING)
                .topic(topic)
                .batchingMaxPublishDelay(Long.MAX_VALUE, TimeUnit.MILLISECONDS)
                .create();

        final int messages = 10;
        for (int i = 0; i < messages; i++) {
            producer.sendAsync("New message - " + i);
        }
        producer.flush();

        final List<MessageId> messageIdList = new ArrayList<>();
        for (int i = 0; i < messages; i++) {
            final MessageId messageId = consumer.receive().getMessageId();
            messageIdList.add(messageId);
            if (i != 4) {
                consumer.acknowledge(messageId);
            }
        }
        consumer.seek(MessageId.earliest);
        Thread.sleep(100);
        consumer.acknowledge(messageIdList.get(4));
        consumer.close();
        consumer = subscribe(topic);
        final Message<String> message = consumer.receive(1, TimeUnit.SECONDS);
        // NOTE: message is null
```

How could you explain to users that you should not acknowledge `messageIdList.get(4)`, otherwise message lost will happen? BTW, if users left one message ID (`messageIdList.get(4)`) to acknowledge after `seek` and they didn't acknowledge other message IDs, what will they expect? Did they expect messages could be received again?

In short, I think we should not handle this corner case. It's more like a "bug" on purpose.

Reply:
Users should not continue to process/ack the messages before the seek operation. It's not the same case as I provided which is normal usage. 

And we should ensure after the seek operation. We should not return the cached messages before the seek operation to users. If users try to cache the messages or just inflight messages, users need to guarantee they will not ack the inflight or cached messages before the seek operation.

> BTW, if users left one message ID (messageIdList.get(4)) to acknowledge after seek and they didn't acknowledge other message IDs, what will they expect? Did they expect messages could be received again?

Yes, they should receive all the messages after the new seek position no matter what happened before the seek operation.

Reply:
> Users should not continue to process/ack the messages before the seek operation.

This implicit rule is never documented and even it's documented, it could be a burden to users.

Assuming there are a message ID set, given the following steps:
1. Acnowledge some message IDs
2. Perform some operations, e.g. `seek`.
3. Acknowledge other message IDs.

```java
consumer.acknowledge(msgId1);
consumer.doSomething(); // pseudo code
consumer.acknowledge(msgId2);
```

The rule means, the steps above could have the different result with acknowledging the whole message ID set:

```java
consumer.acknowledge(msgId1);
consumer.acknowledge(msgId2);
```

A more confusing thing is that with this rule, consumer can acknowledge another message ID that has the same position with `msgId1` to have the same result. i.e.

```java
// CODE 1
consumer.acknowledge(msgId1);
consumer.doSomething();
consumer.acknowledge(msgId2);
consumer.acknowledge(msgId3);
// msgId3.getLedgerId() == msgId1.getLedgerId()
// msgId3.getEntryId() == msgId1.getEntryId()
// msgId3.getBatchIndex() == msgId1.getBatchIndex()
``` 

is equivalent with

```java
// CODE 2
consumer.acknowledge(msgId1);
consumer.acknowledge(msgId2);
```

What's more confusing is, `msgId1.equals(msgId3)` is true and if we replace `consumer.acknowledge(msgId3)` with `consumer.acknowledge(msgId1)` in `CODE 1`, `CODE 1` won't be equivalent with `CODE 2`.

Reply:
The point I insisted on is, the following two code snippets should be equivalent.

```java
        final List<BatchMessageIdImpl> msgIds = new ArrayList<>();
        for (int i = 0; i < numMessages; i++) {
            msgIds.add((BatchMessageIdImpl) consumer.receive().getMessageId());
        }
```

```java
        final List<BatchMessageIdImpl> msgIds = new ArrayList<>();
        for (int i = 0; i < numMessages; i++) {
            final MessageIdImpl messageId = (MessageIdImpl) consumer.receive().getMessageId();
            MessageId deserialized = MessageId.fromByteArray(messageId.toByteArray());
            msgIds.add((BatchMessageIdImpl) deserialized);
        }
```

This PR works well for the assumption above.

Your solution to the corner cases you described is very hacky. You created the **same** (i.e. `equals` returns true) message ID and archive a different goal.

Reply:
> If users try to cache the messages or just inflight messages, users need to guarantee they will not ack the inflight or cached messages before the seek operation.

Just like this rule you've mentioned here, when message IDs from deserializations are used, there is a rule that all message IDs should be acknowledged (for simplicity, it does not count the cumulative ACK case). IMO, if users load a set of message IDs and he only acknowledges partial of them before seeking, it should be an incorrect use.

> It's not the same case as I provided which is normal usage.

As I've explained above, it's not normal usage when you use message IDs from deserializations.

The wrong code users should not write:

```java
var id0 = loadMessageId();
consumer.acknowledge(id0); // batch index: 0, batch size: 2
var id1 = loadMessageId(); // batch index: 1, batch size: 2
// User does not acknowledge id1 before seek
consumer.seek(MessageId.earliest);
// Instead, user acknowledges the outdated id1 after seek.
consumer.acknowledge(id1);
```

The correct code users should write:

```java
var id0 = loadMessageId();
consumer.acknowledge(id0); // batch index: 0, batch size: 2
consumer.seek(MessageId.earliest);
storeMessageId(consumer.receive().getMessageId());  // batch index: 0, batch size: 2
storeMessageId(consumer.receive().getMessageId());  // batch index: 1, batch size: 2
var id1 = loadMessageId(); // batch index: 0, batch size: 2
var id2 = loadMessageId(); // batch index: 1, batch size: 2
consumer.acknowledge(id1);
consumer.acknowledge(id2);
```

----

Regarding the reconnection or ack failure, it's just the same. When `acknowledgeAsync` is called on all message IDs of a batch, the cache will be removed no matter if the acknowledgment succeeded. If only partial message IDs of a batch are acknowledged, when users received messages again, they should persist message IDs again and use the new message IDs, including the repeated positions, instead of the old message IDs.

Reply:
The `id2` can be acked first, but the client crashed while processing message 1 (Shared subscription and users multiple threads to process messages). I mean that users do not want to miss the ack intentionally. But we can't guarantee the messages of a batch will be processed in order of a Shared subscription. Even if it is a Failover or Exclusive subscription, users can also ack individually and process messages out of order.

> Regarding the reconnection or ack failure

I point out this one because a server-side cursor reset can cause the reconnection. But it should not be a problem that this PR should fix.

Reply:
I just thought again. If we have to use completely new `BatchMessageIdImpl` objects in regular cases, i.e. retrieve `MessageId` from `Message`, in the same case when `BatchMessageIdImpl` objects are retrieved from deserializations, the cache should be cleared.